File: LeetCode-Daily/Heap/Zero Array Transformation 3.cpp
/*
* Leetcode : 3362
* T.C : O(N * QlogQ) // Q log Q for past and maxh
* S.C : O(Q)
*/
class Solution {
public:
    int maxRemoval(vector<int>& nums, vector<vector<int>>& queries) {
        sort(queries.begin(), queries.end()); // sort by first element of each query

        priority_queue<int> maxh; // max-heap (stores negative of values to mimic Python's -heapq)
        priority_queue<int, vector<int>, greater<int>> past; // min-heap

        int usedQuery = 0;
        int j = 0;
        int n = nums.size();

        for (int i = 0; i < n; ++i) {
            // Push new queries whose index matches current i
            while (j < queries.size() && queries[j][0] == i) {
                maxh.push(queries[j][1]); // pushing directly, no need to negate
                ++j;
            }

            nums[i] -= past.size();

            // While there are still max operations we can do and they are valid
            while (nums[i] > 0 && !maxh.empty() && maxh.top() >= i) {
                nums[i]--;
                int el = maxh.top(); maxh.pop();
                past.push(el);
                usedQuery++;
            }

            if (nums[i] > 0) return -1;

            // Remove past entries that expire at current i
            while (!past.empty() && past.top() == i) {
                past.pop();
            }
        }

        return queries.size() - usedQuery;
    }
};

File: LeetCode-Daily/Heap/Zero Array Transformation 3.py
### Tle 
class Solution:
    def maxRemoval(self, nums: List[int], queries: List[List[int]]) -> int:
        queries.sort()
        maxh = []
        heapq.heapify(maxh)
        past = []
        usedQuery = 0
        heapq.heapify(past)
        j = 0
        for i in range(len(nums)):
            while j<len(queries) and queries[j][0] == i:
                heapq.heappush(maxh,-queries[j][1])
                j += 1
            nums[i] -= len(past)
            while nums[i] > 0 and maxh and heapq.nsmallest(1,maxh)[0] <= -i:
                nums[i] -= 1
                el = heapq.heappop(maxh)
                heapq.heappush(past,-el)
                usedQuery += 1
                # print(maxh)
            if (nums[i] > 0):
                return -1
            
            while past and heapq.nsmallest(1,past)[0] == i:
                # print(i)
                heapq.heappop(past)
        return len(queries) - usedQuery
            
